mlir/lib/Dialect/MIOpen/Generator/Conv2dGenerator.cpp

```diff
@@ -66,8 +69,125 @@ void strToTokens(const std::string &arguments,
   }
 }
 
+Optional<SmallVector<int64_t, 5>>
+canonicalizeDims(const SmallVector<int64_t, 5> &dims, const StringRef layout,
+                 const StringRef canonicalLayout) {
+  SmallVector<int64_t, 5> ret;
+  for (const char &c : canonicalLayout) {
+    auto location = layout.find(c);
+    if (location == StringRef::npos) {
+      return None;
+    }
+    ret.push_back(dims[location]);
+  }
+  return ret;
+}
+
```

Comment:
I see what you are trying to do here. Now, instead of trying to canonicalize to a specific dimension, why not capture the entire layout - dimension as a map. This way you don't need to populate a hard coded layout, or use hard coded index to find a specific dimension.

Reply:
... yeah, that's a good point. I should probably do it that way instead. I can even use llvm's StringMap. Thank you for noticing a better way to write this function!